### DISCUSSION TITLE
Goal-focused Metrics (CR on all stats)

### DISCUSSION THREAD

**Discussion by ro-savage** [plausible/analytics#1296] [Anything Else]
I would like to click a goal, and then view all my sources at once and see that "CR". This would allow me to see which campaigns are doing the best at meeting a goal, not in absolute terms, but in the percentage terms.

This is very important when trying to decide what campaigns to focus on, and right now you have to check each campaign individually.

Right now you can filter by a "Source" then go to "Goals" and see the percentage of people from that source who reach the goal.

However, the opposite is not possible. If click a goal, I can see how many people from each source hit that goal but I can't see that as percentage of the total amount.

I believe this same "CR" could be applied to every table when a goal filter has been applied. (Although, for most sources is the most important.

Having had a  quick look at the code, I think I can probably put a PR up for this, if it's of interest.

**Comment by metmarkosaric** [comment]
thanks @ro-savage! this would be nice and i'd like to see it myself. there's a thread with this feature request here https://github.com/plausible/analytics/discussions/1115. we'd definitely welcome a PR, it will help speed this up and help us make it live to everyone faster! thanks!

**Reply by ro-savage** [reply]
No worries @metmarkosaric - I've done a PoC but need some guidence.

I think the 'best' way to do this would be to have the API check when a goal filter has been applied, and if it has send back on each endpoint an additional field such as `ConversionRate` that can then just be read directly in the frontend.

However, after banging my head for a while against erlang / elixir (both which i've never used) I wasn't getting very far.

So instead, I went down the route of making two seperate requests on the frontend
1. If Goal filter is applied
2. Request original stats + stats with goal removed
3. Use the goal removed data to calculate the the Conversion Rate and at it to the state.

This strategy means when double the requests to the server when the filter is applied.

I am happy to continue down this path and make it generic enough that it can be easily applied to every stats panel. However, it might not be an acceptable solution to you.

Please let me know your thoughts - here's a quick video of the change.

https://www.loom.com/share/14939791445544d7a497bdf3438b6a14

**Reply by ro-savage** [reply]
Here's a quick and dirty implemention so you understand what I mean.

```js
  fetchReferrers() {
    if (this.hasGoalFilter()){
      const fetchPromise = api.get(`/api/stats/${encodeURIComponent(this.props.site.domain)}/sources`, this.props.query, {show_noref: this.showNoRef()})

      const queryWithoutGoal = {
        ...this.props.query,
        filters: { ...this.props.query.filters }
      }
      queryWithoutGoal.filters.goal = null
      const fetchWithoutGoalPromise = api.get(`/api/stats/${encodeURIComponent(this.props.site.domain)}/sources`, queryWithoutGoal, {show_noref: this.showNoRef()})

      Promise.all([fetchPromise, fetchWithoutGoalPromise]).then(values => {
        const referrersWithGoal = values[0]
        const referrersWithoutGoal = values[1]
        const referrers = referrersWithGoal.map(referrer => {
          const allVisitors = referrersWithoutGoal.find(ref => ref.name === referrer.name).count
          const conversionRate =  referrer.count / allVisitors * 100

          return {
            ...referrer,
            conversionRate,
          }
        })
        this.setState({loading: false, referrers })
      })
    } else {
      api.get(`/api/stats/${encodeURIComponent(this.props.site.domain)}/sources`, this.props.query, {show_noref: this.showNoRef()})
         .then((res) => this.setState({loading: false, referrers: res}))
    }
  }
```

**Reply by ro-savage** [reply]
Tagging @ukutaht since it looks like he might be the gate keeper here :)

**Reply by ukutaht** [reply]
@ro-savage Thanks for your work on this.

The issue with this implementation is that you don't always get the same referrers in the report. With synthetic data it probably works OK but in the real world your top 9 referrers might not exactly line up with your top 9 referrers for a goal. So in some cases the `allVisitors` cannot be calculated. This issue gets worse the more referrers you have.

We had the same issue with some other reports and we had to break them into 2 sequential queries. First to get the top 9 referrers for a goal. And a second one to get `allVisitors` number for all of these referrers (this is kind of how time on page is calculated in the 'Top pages' report)

Now, to be able to do this you need a `OR` filter or something similar so you can query with a filter like `referrer=ref1 OR ref2 OR ref3`. This is possible on the backend and through the API but currently not through the React frontend.

This is all to say that you probably need to get on the backend to write a 'proper' query that always works in the real world. Or wait until the filtering system is overhauled to allow `OR` filters.

**Reply by ro-savage** [reply]
Thanks for the feedback. That certainly make sense, my first attempt was trying to build something based off the `fetch_top_stats` function to add to the other stats endpoints.

Unfortunately I think that the erlang stuff is going to outside of my capabilities. However, if you point me in the direction of a PR that adds new metrics I could give it a go.

Otherwise, happy to contribute any help to the react side of things to implement thing.

BTW Awesome work both of you - I'd be looking for something like this for a long time - really glad to find it and see the success you're having. Well deserved!

**Reply by ro-savage** [reply]
I've put something together that retrieve the stats server side. But its probably pretty horribly done.

I'll try clean it up a little and push it as PR for feedback when I get the chance.

**Comment by metmarkosaric** [comment]
this is now live thanks to the contribution by @ro-savage! conversion rate (CR) for all the metrics is shown both in the main view and in the details view.
